[CMake] Use separate namespace for demangling symbols used by runtime

edymtt opened the pull request:
This matches the behavior introduced with #30733 -- quoting the
explanation here for the sake of convenience:
    
> Since libDemangling is included in the Swift standard library,
> ODR violations can occur on platforms that allow statically
> linking stdlib if Swift code is linked with other compiler
> libraries that also transitively pull in libDemangling, and if
> the stdlib version and compiler version do not match exactly
> (even down to commit drift between releases). This lets the
> runtime conditionally segregate its copies of the libDemangling
> symbols from those in the compiler using an inline namespace
> without affecting usage throughout source.

Addresses rdar://142550635

edymtt commented:
@swift-ci please smoke test

etcwilde commented on Runtimes/Core/CMakeLists.txt:
```diff
@@ -124,6 +124,11 @@ add_compile_definitions(
+  # the inline namespace to distinguish symbols from those built for the
+  # compiler, in order to avoid possible ODR violations if both are statically
+  # linked into the same binary. (see also commit message for 5b1daa9055c99904c84862ecc313641fd9b26e63)
+  $<$<COMPILE_LANGUAGE:C,CXX>:-DSWIFT_INLINE_NAMESPACE=__runtime> # Demangling, runtime, Concurrency, RemoteInspection, SwiftRemoteMirror, swift-reflection-test
```
Since this is only needed by things that pull in `libDemangling`, we should make this a `PUBLIC` define coming from the libDemangling library. Then anything that links Demangling will pick up the appropriate setting.

edymtt replied:
Thanks for the guidance -- I admit I blindly matched the existing code, without seeing its underlying purpose and translating it to proper modern CMake patterns.

etcwilde replied:
This global stash of definitions should eventually have the definition attached to the appropriate libraries. This was mostly for expedience when deciphering the logic in `add_swift_target_library` was taking too long.

etcwilde commented on Runtimes/Core/CMakeLists.txt:
```diff
@@ -124,6 +124,11 @@ add_compile_definitions(
+  # the inline namespace to distinguish symbols from those built for the
+  # compiler, in order to avoid possible ODR violations if both are statically
+  # linked into the same binary. (see also commit message for 5b1daa9055c99904c84862ecc313641fd9b26e63)
+  $<$<COMPILE_LANGUAGE:C,CXX>:-DSWIFT_INLINE_NAMESPACE=__runtime> # Demangling, runtime, Concurrency, RemoteInspection, SwiftRemoteMirror, swift-reflection-test
```
LLVMSupport explicitly uses `inline namespace __swift { inline namespace __runtime {` instead of using the `SWIFT_BEGIN_INLINE_NAMESPACE` and `SWIFT_END_INLINE_NAMESPACE` macros from `NamespaceMacros.h`. Demangling is built as part of both the compiler and runtime so it needs a switch. We should only need this on the libDemangling library.

edymtt committed:
[CMake] Use separate namespace for demangling symbols used by runtime

This matches the behavior introduced with #30733 -- quoting the
explanation here for the sake of convenience:

> Since libDemangling is included in the Swift standard library,
> ODR violations can occur on platforms that allow statically
> linking stdlib if Swift code is linked with other compiler
> libraries that also transitively pull in libDemangling, and if
> the stdlib version and compiler version do not match exactly
> (even down to commit drift between releases). This lets the
> runtime conditionally segregate its copies of the libDemangling
> symbols from those in the compiler using an inline namespace
> without affecting usage throughout source.

Addresses rdar://142550635

etcwilde commented on Runtimes/Core/runtime/CMakeLists.txt:
```diff
@@ -95,7 +95,8 @@ target_include_directories(swiftRuntime PRIVATE
   $<$<PLATFORM_ID:Windows>:User32>
-  swiftShims)
+  swiftShims
+  swiftDemangling)
```
We shouldn't need this. I don't see `SWIFT_INLINE_NAMESPACE` getting used in `swiftRuntime`.

edymtt replied:
It is used indirectly since a few sources in `runtime` include Demangling headers, which in turn include `namespaceMacros.h` -- if I don't add that line, I get linker errors because we are not able to find symbols like `mangleNode`.

etcwilde replied:
Ah, good catch.
